### PR TITLE
[DeepSeek] Enable checkpoint load from HF

### PR DESCRIPTION
**Pull request by kwen2501**
Enable loading weights from HF checkpoint.

1. Download
- Added `download.py` to allow user download a HF checkpoint into local disk cache.
- Usage: `python download.py {model_id}`

2. Load weights
- Added `checkpoint.py` to load tensors from HF cache dir into a model.
- The model can be a model chunk or full model.

3. Various code refactor
- Moved `ModelArgs` into a separate file, adding DeepSeek config registry

4. Added support for DeepSeek-V2
- Greedy routing
- Softmax score function

5. Added example run.py based on [DeepSeek-V2-Lite](https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite/tree/main), a 16B toy MoE.
`torchrun --standalone --nproc-per-node 4 run.py`

### CODE REVIEW

File: torchtitan/experiments/deepseek_v3/run.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# torchrun --standalone --nproc-per-node 4 run.py
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/run.py:7]: is this file named 'run' instead of 'train' because it's not supposed to train?  It looks like it is using a training oriented pipeline-schedule (not fwd-only) but missing a loss function etc.  Maybe its just a temporary file for bootstrapping...

File: torchtitan/experiments/deepseek_v3/model_config.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (wconstab) [inline torchtitan/experiments/deepseek_v3/model_config.py:1]: nit: model_args.py? since ModelArgs

**Review by wconstab — APPROVED** [review]

File: torchtitan/experiments/deepseek_v3/run.py (hunk 1 above)
Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/run.py:7]: It is a temp file for bootstrapping. And it will diverge into `generate.py` and `train.py` later.

File: torchtitan/experiments/deepseek_v3/model_config.py (hunk 2 above)
Comment (kwen2501) [reply torchtitan/experiments/deepseek_v3/model_config.py:1]: Hmm, I think both work. My purpose of creating this file is to store configs of different DeepSeek flavors, like the `DeepSeek-V2-Lite` in it.

File: torchtitan/experiments/deepseek_v3/run.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    # Run forward
+    if pp_rank == 0:
+        y = pp_schedule.step(x)
+    else:
+        y = pp_schedule.step()
+
+    if pp_rank == pp_size - 1:
+        print(y.shape)
+
```
Comment (lessw2020) [inline torchtitan/experiments/deepseek_v3/run.py:80]: maybe a simple "Success - forward completed" or similar print just to show the forward completed properly?

**Review by lessw2020 — APPROVED** [review]
Looks great - also nice to see the Lite version being supported as that will be better for single node work. 
Also the SPMD weight loading logic is superb... :)   Happy to see we are getting some re-use out of that previous work.
